File: client/src/game/events.ts
import {EVENT_RING} from '../sim/wasm'

/**
 * Firing effects without firing them eight times.
 *
 * **The classic rollback bug is one hit sound playing once per replay.** The
 * sim avoids it by setting flags in state rather than firing anything (see
 * sim/events.go); this is the other half — the view fires effects only for
 * frames that can no longer be simulated again, and never fires a frame twice.
 *
 * Two rules and nothing else:
 *   1. Only frames at or below the confirmed line, because a predicted frame
 *      can still be rolled back and its events replaced.
 *   2. Only frames not already fired, because the confirmed line is read every
 *      display frame and mostly has not moved.
 */
export interface Fired {
  frame: number
  /** 0 or 1 — the player the effect happened *to*, which is where to draw it. */
  seat: number
  /** The EVENT_* bits set on that seat for that frame. */
  bits: number
}

export interface Events {
  /** Fires everything settled since the last call, oldest frame first. */
  drain(confirmed: number, fire: (f: Fired) => void): void
  /** Forgets what has been fired. For a new match, which starts at frame 0. */
  reset(): void
}

/**
 * @param read the sim's per-frame event words, one per seat. Called only for
 * frames inside the sim's own ring; older ones read as zero there anyway,
 * which is why that ring is where the catch-up below stops.
 */
export function createEvents(read: (frame: number) => readonly [number, number]): Events {
  let firedThrough = -1

  return {
    drain(confirmed, fire) {
      // A backgrounded tab can confirm hundreds of frames at once. The sim has
      // long since overwritten them, so this is where the loop stops rather
      // than a decision about how many effects are too many.
      const from = Math.max(firedThrough + 1, confirmed - EVENT_RING + 1)

      for (let f = from; f <= confirmed; f++) {
        const bits = read(f)
        for (let seat = 0; seat < bits.length; seat++) {
          if (bits[seat] !== 0) fire({frame: f, seat, bits: bits[seat]})
        }
      }
      if (confirmed > firedThrough) firedThrough = confirmed
    },

    reset() {
      firedThrough = -1
    },
  }
}

File: sim/events.go
package sim

// Rollback-safe events (02 Architecture/Rollback Netcode.md, 04 Art/Audio.md).
//
// **The classic rollback bug is a hit sound playing eight times.** It happens
// the obvious way: the sim fires an effect imperatively the moment a hit lands,
// a packet arrives, the frame is rewound and replayed, and the effect fires
// again on every replay. A frame that is rolled back eight times is simulated
// nine times, and the player hears nine punches for one hit.
//
// The fix is the one the design note names, and it is structural rather than
// careful: **the sim sets flags in state and fires nothing.** The flags roll
// back with everything else, so a replayed frame recomputes them rather than
// accumulating them. The view then fires effects only for flags on *confirmed*
// frames — frames whose inputs are all known and which therefore cannot be
// simulated again.
//
// Built in M2 with nothing attached to it, deliberately. There are no sounds
// and no particles yet; retrofitting this once there are means auditing every
// effect in the game, and the bug it prevents is silent, intermittent and only
// reproducible online.
//
// The flags live on the player rather than on the match because almost every
// effect has a position: a hit spark belongs where the fighter is standing.
const (
	// EventHit is a clean hit taken — the strike connected and was not
	// blocked. On the *defender*, like every flag here: effects are drawn on
	// the thing they happened to.
	EventHit int32 = 1 << 0

	// EventBlock is a hit blocked.
	EventBlock int32 = 1 << 1

	// EventThrown is a throw connecting, on the player being thrown.
	EventThrown int32 = 1 << 2

	// EventKnockdown is hitting the floor.
	EventKnockdown int32 = 1 << 3

	// EventSuper is a super activating, on the player who spent the bars. The
	// activation flash the design note wants is this flag plus a view that
	// draws something; the *freeze* it also asks for is a gameplay pause and
	// is not built (see D78).
	EventSuper int32 = 1 << 4
)

// clearEvents wipes both players' flags. Called at the top of every frame, so
// a flag means "this happened on this frame" and never "this happened at some
// point recently" — a flag that outlived its frame would be fired again on
// every confirmed frame it survived into, which is the bug in a slower form.
func (s *GameState) clearEvents() {
	s.Players[0].Events = 0
	s.Players[1].Events = 0
}

// packEvents is both players' flags in one word, player 0 low. What crosses the
// WASM boundary, where a call returning two numbers costs two calls.
func (s *GameState) packEvents() uint32 {
	return uint32(uint16(s.Players[0].Events)) | uint32(uint16(s.Players[1].Events))<<16
}
